src/hooks/useImmediateState.ts

```diff
@@ -0,0 +1,33 @@
+import { useRef, useEffect, useState, useCallback } from 'react';
+
+type SetStateAction<S> = S | ((prevState: S) => S);
+type SetState<S> = (action: SetStateAction<S>) => void;
+
+/** This is a drop-in replacement for useState, limited to object-based state. During initial mount it will mutably update the state, instead of scheduling a React update using setState */
+export const useImmediateState = <S extends {}>(init: S): [S, SetState<S>] => {
+  const isMounted = useRef(false);
+  const initialState = useRef<S>({ ...init });
```

Comment:
Just had a thought about this - we aren't using this object anywhere other than initial render so the use of useRef might be misleading/unnecessary.

Reply:
We are using this as a throwaway target for unmounted setState calls but otherwise yea, not really needed on client-side.

But during React server-side rendering this can actually matter, since some calls will reset component rendering and start over, which means that since our client is external we might get subtly different results.

It probably doesn't matter but that's why I chose to use refs.